### ISSUE TITLE
Invisible snipers improvements

### ISSUE THREAD

**Issue by GoogleFrog** [ZeroK-RTS/Zero-K#2119]
Sniper shots should be invisible. We can possibly poison the callin to filter them out. An engine change may be best. @sprunk you bought this to a head so it's your job.

**Comment by Anarchid** [comment]
The alternative is making "sniper approximately reveals its position when firing" a rule.

**Comment by GoogleFrog** [comment]
Projectile tracking is not approximate though.

**Comment by Anarchid** [comment]
Due to LoS, it is, though i guess a line segment isn't too inaccurate either.

**Comment by sprunk** [comment]
I think the current situation is fine and disagree with making it invisible to lua unless it is also made invisible to humans.

I think the marker currently serves a role similar to unit tactical AI:
 * does not do anything a human can't technically do. The bullets are visible and follow a straight line path.
 * approximates higher level play. Pros know to look for the bullet and "feel" where the sniper probably is.
 * makes sure #decisions == #actions. One always wants to look for sniper bullets so there are 0 decisions. Without the marker there is an action to do (actively look in the area).
 * diminishes punishment for not paying attention without completely removing it.

I don't think there's even a practical problem:
 * dealing with sniper still has a lot of depth in it. How much to rely on the marker (the sniper might've moved since), and what to do at all (attack, potentially into defense/riots/etc? hold the position but take attrition? retreat but cede ground?).
 * if sniper got noticeably weaker (and if that's considered a problem given recent glaive/rocko buffs) it can be buffed. Making the projectile invisible to lua AND humans would be the most quant and I could get behind it.

**Comment by Anarchid** [comment]
I think my Spectres made cost in the game which likely provoked this ticket. Even if only by causing deaths of Halberds sent to hunt for them inside swirling masses of Glaives and Warriors, rather than by direct kills.

**Comment by Aquanim** [comment]
I dislike that this widget exists. I would also dislike sniper shots being invisible. Snipers revealing their position on firing is not great but I object to that less.

At this time I do not have an argument for those dislikes which is based on the design principles of ZK. My dislike stands.

**Comment by GoogleFrog** [comment]
The current situation is awful, the worst of the options. It is the situation in which players gain advantage by knowing to use an unsupported widget (hence it is relatively unknown and obscure) and with poor unit AI. The widget is not being supported because it is ugly in design, implementation and graphically.

The sniper widget is ugly graphically simply because no widget should be drawing lines on the map with Spring.MarkerAddLine. This just looks really ugly and persists on the map. Any implementation of sniper shot visibility aids that wish to be included must look decent to be considered supported.

The sniper widget is wasteful in the sense that all it does is fully subvert another game mechanic. Tactical AI is nothing like the sniper widget in this sense. Tactical AI does not undermine simulated projectiles, it interacts with simulated projectiles to create particular interactions. The sniper widget does not create interactions, it just removes a mechanic, and it does so expensively by polling every projectile every frame. Here are some potential better implementations:
 * Just make the projectile more visible. There is sure to be a projectile type and configuration which is only visible within LOS and that has a fade out time. This deals with the graphics issue as well.
 * Have LuaRules send the position and direction of sniper shots when they enter LOS. Gadgets are in a better position to track this.

These solutions should not be implemented because they are trumped by the greater design issue. In the majority of cases a visible sniper shot will lead players to have high confidence about the position of enemy snipers. They would be so confident that the failure of their units to shoot at that location would be a unit AI issue. In other words, a visible sniper shot is basically equivalent to snipers decloaking when they fire. You were applying the design principals locally without thinking it through to its conclusion.

Sniper was not perfect. It has design issues. However, they were minor issues which are unimportant at the moment, on the same order as Solar armour, LLT firing cost and Defender overshoot (although this one is being forced now too). It's a minor problem which was unexploited in the current playerbase so it could be left for later. You didn't open a ticket and say "Lets put some work in and fix this dangling design issue". You forced the issue by opening pandoras box and destroying our ability to get back to the old state of affairs. The new state is worse and I don't want to be forced to do anything about it.

> I think the current situation is fine and disagree with making it invisible to lua unless it is also made invisible to humans.

Did I not say that sniper shots should be invisible? There is difficulty here though. Players should be able to see their own sniper shots because it is good feedback for showing which of their snipers just shot. I thought that shots are currently sufficiently visible for players to see their own snipers but not visible enough to be consistently spotted by opponents. Ideally I'd like to retain visibility to the owner of the sniper but it may not be possible now.

**Comment by Licho1** [comment]
Sniper shots being invisible to humans is horrible and I would be strongly against it.
Sniper being invisible to lua is ok solution.

We cannot be too dogmatic about some things. 
Human tracking bertha shot is ok and fun. 
Widget triangulating exact position of BB within 5 elmo after 3 shots is bad and takes the fun away too - even if it is official.

We could either block all custom widgets or scramble shot data. 
Scrambling data sounds like a better solution to me for now.

**Comment by Anarchid** [comment]
> Just make the projectile more visible. There is sure to be a projectile type and configuration which is only visible within LOS and that has a fade out time. This deals with the graphics issue as well.

Lasercannon with reasonably long-lasting cegtrail?

**Comment by sprunk** [comment]
7a241f018ff6284407d29721e9459eedcfc3f516 - invisible to widgets

84602940c72003d63e82ee7ad8911f685314bef7 - invisible to enemy humans (own allyteam can still see the bullets)

**Comment by sprunk** [comment]
Things left to do:
 [ ] make the engine send Projectile{Created,Destroyed} to unsynced gadgets for performance gain
 [ ] either get a non-crappy bullet model OR make the engine call DrawProjectile for sprites and reinstate the old bullet

**Comment by Anarchid** [comment]
9f74c3544dfea91a08b8293af0876ab0165c2e4a
![image](https://cloud.githubusercontent.com/assets/3822768/24582484/87d30c78-1739-11e7-837c-b0806f8e841d.png)

**Comment by Anarchid** [comment]
I assume the shadow can be killed using some kind of draw override as well. Perhaps the rendering can even be made additive.

**Comment by sprunk** [comment]
Unsure if less ugly

**Comment by Anarchid** [comment]
Could be downsized, looks reasonably well in practice and not that different from the original.
Another way to get the purple blob back is to  actually just draw it with Lua.

**Comment by GoogleFrog** [comment]
Pay attention, the projectiles did not produce light earlier as well. The projectile should look cloaked and this model does not look less ugly.

**Comment by Anarchid** [comment]
So, something like an actual physical cloaked thing? I could try bolting a cloak shader on an arrow. Though it seems @Licho wants human visible snipers rather than perfectly invisible ones.

**Comment by Licho1** [comment]
Yes humans need to see them. Visible shot is ok, decloak is ok, seismic indicator is ok (for me, I know google frog doe snot like it due to apparent unit stupidity)

**Comment by GoogleFrog** [comment]
Hold on, decloak is ok? Is there a way to make that a reasonable change to the unit though?

**Comment by Licho1** [comment]
You could add brief radar only decloak (then even hlt might miss) and if it makes sniper too weak, perhaps we can break quant rule and add more paper to its paper armor.

**Comment by Anarchid** [comment]
Spectre [https://github.com/ZeroK-RTS/Zero-K/commits/master/units/armsnipe.lua](used to be) faster back when it was called Sharpshooter. Decloak-on-fire will make its relocation after shot a big part of the hide and seek. So perhaps buff (un-nerf) speed?

Decloak on fire is the cleanest solution because it *removes* an exception.  It's been long since things were *removed*.

**Comment by sprunk** [comment]
Decloak on fire is un-quant but I like the rule exception removal and it's better design-wise (no tedium).

My preference is
1) completely invisible - quants and no tedium but Licho veto
2) decloak on fire - no tedium but un-quant
3) cloak yet visible bullet - status quo but tedium

**Comment by GoogleFrog** [comment]
There are more options. A good one come up with recently is to make the projectile invisible but to make the explosion effect point approximately back towards where the sniper was. You're supposed to get the general direction of sniper fire.

**Comment by Licho1** [comment]
Can it be done without much hackery?

**Comment by Anarchid** [comment]
Origin-pointing explosion effects is just hacking together some directional CEG. Aside from CEG being a bastard child of regex and brainfuck, no hackery. The indication of which sniper fired can be an animation (barrel jerking etc).

**Comment by sprunk** [comment]
The indication of which sniper fired would be the bullet (visible for owner).

Directional CEG sounds good, as long as it lingers for a few seconds (else tedium).

**Comment by Anarchid** [comment]
![image](https://cloud.githubusercontent.com/assets/3822768/24726523/5cbc4bec-1a5b-11e7-9a54-34abad864f2f.png)
https://github.com/ZeroK-RTS/Zero-K/commit/b51726a65a9103b9169f07052cd8d685d270ad6d

**Comment by sprunk** [comment]
The bullet is still visible for the enemy. First I need to PR the engine to make DrawProjectile able to block drawing sprites. This is needed because the "use model" solution results in ugly bullets and the "remove altogether" results in no feedback for the owner.

**Comment by sprunk** [comment]
Doable without touching engine. The projectile can use a fake model for hiding purposes and the sprite can be drawn by LuaGL for the owner team (and possibly the enemy team as well when approaching the target, as suggested earlier).

**Comment by Anarchid** [comment]
This seems obsolete?

**Comment by sprunk** [comment]
Making allies (but not enemies) see the bullet with greater visibility would still be good.

**Comment by GoogleFrog** [comment]
Idk about that. Seems done to me.
